### PR TITLE
Handle pagination errors in chats

### PR DESCRIPTION
**Pull request by janicduplessis**
### Details

Currently we do not handle errors properly in `getOlderActions` and `getNewerActions`. All we do is set the loading state back to false, but that will cause the list `onEndReached` or `onStartReached` to be called again and a new network request to load more will be triggered. If we are near any edge of the chat, or even worse if we have a chat with few messages we will trigger these requests in a loop.

To solve this we should add an error state to those pagination methods and display an error UI with an option to retry loading. This will prevent this request loop from happening.

Here's the UI I implemented for this error state

Top error:

image

Bottom error:

image

Example with comment linking:

https://github.com/Expensify/App/assets/2677334/1c281a27-91ba-490f-b39a-866deb5f5f51

### Fixed Issues
$ https://github.com/Expensify/App/issues/40641
PROPOSAL:

### Tests

- [x] Verify that no errors appear in the JS console
- [x] Test error state by failing pagination requests adding the following code in `HttpUtils.ts` [here](https://github.com/janicduplessis/Expensify-App/blob/16fa023201721525aadbd28e9975b0df53bddb80/src/libs/HttpUtils.ts#L108-L109)

```ts
if (url.includes('GetNewerActions') || url.includes('GetOlderActions')) {
    return {jsonCode: CONST.JSON_CODE.EXP_ERROR};
}
```

- [x] Test comment linking

### QA Steps

- [ ] Verify that no errors appear in the JS console
- [ ] Test error state by failing pagination requests adding the following code in `HttpUtils.ts` [here](https://github.com/janicduplessis/Expensify-App/blob/16fa023201721525aadbd28e9975b0df53bddb80/src/libs/HttpUtils.ts#L108-L109). It might also be possible to simulate this by turning off internet after loading a chat.

```ts
if (url.includes('GetNewerActions') || url.includes('GetOlderActions')) {
    return {jsonCode: CONST.JSON_CODE.EXP_ERROR};
}
```

- [ ] Test comment linking

1. Go in a long chat and scroll to an old message
2. Right click and "Copy link"
3. Go in another chat and paste the link
4. Clear the cache (Account Settings -> Troubleshoot -> Clear cache and restart)
5. Go back to the chat with the link and click it
6. You will be in the middle of a chat and can test loading older / newer messages

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/2677334/1c281a27-91ba-490f-b39a-866deb5f5f51

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

image

image

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

File: src/pages/home/report/ReportActionsView.tsx [hunk 1]
```diff
@@ -297,53 +307,92 @@ function ReportActionsView({
...
-            Report.getOlderActions(reportID, oldestReportAction.reportActionID);
-        }
-    }, [network.isOffline, isLoadingOlderReportActions, isLoadingInitialReportActions, oldestReportAction, hasCreatedAction, reportID, reportActionIDMap, transactionThreadReport]);
+            if (!isEmptyObject(transactionThreadReport)) {
+                // Get newer actions based on the newest reportAction for the current report
+                const oldestActionCurrentReport = reportActionIDMap.findLast((item) => item.reportID === reportID);
+                Report.getNewerActions(oldestActionCurrentReport?.reportID ?? '0', oldestActionCurrentReport?.reportActionID ?? '0');
 
-    const loadNewerChats = useCallback(() => {
-        if (isLoadingInitialReportActions || isLoadingOlderReportActions || network.isOffline || newestReportAction.pendingAction === CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE) {
```
Comment (janicduplessis) [inline src/pages/home/report/ReportActionsView.tsx:326]: I noticed this was checking `isLoadingOlderReportActions` which I assume was an error since this is `loadNewerChats`, but maybe someone can confirm that this change is ok.

**Comment by janicduplessis** [comment]
Is someone familiar with [this code](https://github.com/janicduplessis/Expensify-App/blob/16fa023201721525aadbd28e9975b0df53bddb80/src/pages/home/report/ReportActionsList.tsx#L563)? I noticed the loading indicator for newer messages never seem to show even when comment linking in the middle of a chat that has new messages to load. This also causes the error state not to show since I'm implementing it in `ListBoundaryLoader`, which I think is the ideal place.

**Comment by janicduplessis** [comment]
@mountiny Do you know if there is an issue related to this already? Should I create one?

This is currently coming from [this convo](https://expensify.slack.com/archives/C05LX9D6E07/p1713454764509159) on slack.

**Comment by janicduplessis** [comment]
Might also need help here with design and texts, not sure who to tag and I can't seem to add the Design label.

**Comment by mountiny** [comment]
@janicduplessis I dont think there is one, feel free to create it please.

Then we can add design label there too to ensure the looks are as we expect

**Comment by janicduplessis** [comment]
@mountiny Thanks! Done https://github.com/Expensify/App/pull/40610

**Comment by mountiny** [comment]
@dannymcclain would you be able to help with the design on this PR? thanks!

**Comment by dannymcclain** [comment]
Going to cc @Expensify/design in here as well. How do you all feel about the `Couldn't load` UI?

A couple thoughts from me:
- I think the message text should be `text-supporting` color
- I think I would propose swapping the button for a spinner when you tap `Try again` instead of swapping out the whole message for the wall of skeleton loaders—it's kinda weird when the whole message disappears, a bunch of skeleton loaders show up, and then it comes back.

image

I realize this is unlikely to happen "in real life" as much as it during me testing this, but still, I think it's pretty weird. I would rather the `Try again` button just change to a spinner, and then if we still can't load, it changes back to the button. If we CAN load new messages, the new messages just replace the whole error message block.

https://github.com/Expensify/App/assets/7516624/d2f707d0-5fe2-4535-8caf-e8ab7aac791c

**Comment by janicduplessis** [comment]
@dannymcclain Thanks, I like the suggestion, will look at making those changes this week.

**Comment by shawnborton** [comment]
Maybe said another way, why exactly does the "Try again" button do to fix the pagination error? Why wouldn't we just do that automatically for the user instead of making them tap a button?

**Comment by janicduplessis** [comment]
The network requests can fail for various reasons, whether it be a network issue, or some outage on our end. This happened recently to both @mountiny and @quinthar, with the current behaviour the requests are retried in a loop so this is why I suggested introducing an error ui. I’m pretty sure I’ve seen similar in other chat apps before.

**Comment by dannymcclain** [comment]
>with the current behaviour the requests are retried in a loop

@shawnborton This is shown in [the attached video here](https://github.com/Expensify/App/issues/40641#issue-2254632660). I've been seeing it a lot in the product too—it's super disorienting and there's seemingly no way to make it stop other than quitting the app and reopening it. So this `Try again` UI feels like a big improvement over those crazy flashing skeletons.

**Comment by shawnborton** [comment]
Yeah that's fair, but why does the Try again button work and the automatic retry doesn't?

**Comment by janicduplessis** [comment]
The main goal is to let the user know there is a problem, and avoid spamming our servers with network requests. Currently it just keeps flashing the loading skeleton while immediately retrying the request. If we show some ui it stops the loop and warn the use about the problem. They can decide to try again and if it fails again then it will show the error state again.

**Comment by janicduplessis** [comment]
It would be possible to retry the requests a few times, but at some point we need to stop if it keeps failing and let the user know. That is the main goal of this change. If we want to add some automatic retry behaviour it should probably be implemented somewhere else though, more at the networking layer.

**Comment by shawnborton** [comment]
So just to make sure I understand correctly, in this PR, if the request fails we'll show you a button to retry. But when you retry, it's likely just going to fail again?

**Comment by janicduplessis** [comment]
Yes, depending on what the problem was, if it was a random network error it might succeed when retrying, but if it is an outage it probably won’t. At least the user will be warned that an error occurred instead of being stuck in a loading loop.

**Comment by janicduplessis** [comment]
The current behaviour that causes a loading loop can also be very bad in case of an outage where all clients will start spamming the backend with network requests.

**Comment by shawnborton** [comment]
Cool, thanks for explaining. I definitely don't want to block on progress here and agree that this is better than what we currently experience. Thanks for hearing me out!

**Comment by janicduplessis** [comment]
There is definitely a good case for retrying failed requests automatically, but that would be a different project. It would be more like whenever a request fail we can try to see what the reason for the failure is and if we think it could work if trying again (let’s say its a network error) then we can re-send the request automatically. Those retries could be limited to only a few times and possibly with a delay. Then in case the retries fail it would throw an error which would be handled by the error states of the app (including the one added here).

**Comment by Gonals** [comment]
@janicduplessis, let me know when this is ready for review!

**Comment by janicduplessis** [comment]
@Gonals Should be ready today, working on it now!

**Comment by janicduplessis** [comment]
Updated the UI as per @dannymcclain suggestions. Here's an updated video of loading failing.

https://github.com/Expensify/App/assets/2677334/82d0a8b7-cb54-4cb1-8a56-f6b658b2406c

File: src/pages/home/report/ReportActionsList.tsx [hunk 2]
```diff
@@ -621,7 +635,11 @@ function ReportActionsList({
         [onContentSizeChange],
     );
 
-    const listHeaderComponent = useCallback(() => {
+    const retryLoadNewerChatsError = useCallback(() => {
+        loadNewerChats(true);
+    }, [loadNewerChats]);
+
+    const listHeaderComponent = useMemo(() => {
```
Comment (janicduplessis) [inline src/pages/home/report/ReportActionsList.tsx:642]: I changed this so we pass a react element instead of a function. When passing a function it causes the component to be unmounted and remounted every time one of the dependencies of useCallback changes. This caused flickers in the loading component as well as resetting any internal state so this fixes it.

File: src/pages/home/report/ReportActionsView.tsx [hunk 3]
```diff
@@ -299,76 +309,116 @@ function ReportActionsView({
...
+                    newestReportAction: newestReportAction.pendingAction,
+                    firstReportActionID: newestReportAction?.reportActionID,
+                    isLoadingOlderReportsFirstNeeded,
+                })}`,
+            );
+
+            if (
+                !force &&
+                (isLoadingInitialReportActions ||
+                    isLoadingNewerReportActions ||
```
Comment (janicduplessis) [inline src/pages/home/report/ReportActionsView.tsx:392]: This used to check `isLoadingOlderReportActions` which I'm pretty sure is an error so I changed it to `isLoadingNewerReportActions`.

**Comment by janicduplessis** [comment]
@Gonals This is ready!

**Comment by Gonals** [comment]
Conflicts! 😆

**Comment by janicduplessis** [comment]
@Gonals Fixed!

**Review by Gonals — DISMISSED** [review]

**Review by Gonals — COMMENTED** [review]
Why don't we have a C+ reviewer for this PR?

**Review by dannymcclain — DISMISSED** [review]
From a design perspective this looks good. What needs to happen to get this over the finish line?

**Comment by janicduplessis** [comment]
Looks like we have some conflicts now, I will fix it soon. Do we want a 2nd review? I will try to get more attention on this.

**Comment by Gonals** [comment]
Assigning @rojiphil as the C+ reviewer. @rojiphil, will you be able to take a look at this?

**Comment by rojiphil** [comment]
> Sure. Will take a look at this today and update.

I am still reviewing but no updates yet. I will continue tomorrow and update again.

**Comment by janicduplessis** [comment]
Conflicts fixed

**Comment by rojiphil** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>MacOS: Chrome / Safari</summary>

TODO

</details>

<details>
<summary>Android: Native</summary>

TODO

</details>

<details>
<summary>Android: mWeb Chrome</summary>

TODO

</details>

<details>
<summary>iOS: Native</summary>

TODO

</details>

<details>
<summary>iOS: mWeb Safari</summary>

TODO

</details>

<details>
<summary>MacOS: Desktop</summary>

TODO

</details>

**Review by rojiphil — COMMENTED** [review]
I have left few comments for your consideration. The overall code looks good though.

**Comment by rojiphil** [comment]
@janicduplessis  Is the test case for  `Testing comment linking`  added to address an edge case, or did we use this for testing error message for newer actions? If this is to address an edge case, please provide testing steps for this as I am not clear on how to reproduce this.

**Comment by janicduplessis** [comment]
Yes, it is to test GetNewerActions, will provide clearer testing steps on how to test comment linking.

File: src/pages/home/report/ReportActionsList.tsx [hunk 4]
```diff
@@ -622,7 +636,11 @@ function ReportActionsList({
...
         if (!canShowHeader) {
             hasHeaderRendered.current = true;
             return null;
```
[the remark this answers is not in the document]
Comment (rojiphil) [reply src/pages/home/report/ReportActionsList.tsx:646]: > am curious about what this canShowHeader is for, meant to raise this in the review but probably forgot.

I think the same reasoning for `listFooterComponent` [here](https://github.com/Expensify/App/blob/e0b6a06d2f34f35cda96ccc4d05252ede111c732/src/pages/home/report/ReportActionsList.tsx#L594) holds good for `listHeaderComponent`. But, `canShowHeader` also considers `scrollingVerticalOffset` [here](https://github.com/Expensify/App/blob/e0b6a06d2f34f35cda96ccc4d05252ede111c732/src/pages/home/report/ReportActionsList.tsx#L584) which seems to make this more complicated. 

> Maybe I should just ignore canShowHeader if there’s an error?

Hmm. Interesting. If there is an error then we may as well show the error UI even during first render. 
Also, we may want to do the same thing for `ListFooterComponent` [here](https://github.com/Expensify/App/blob/e0b6a06d2f34f35cda96ccc4d05252ede111c732/src/pages/home/report/ReportActionsList.tsx#L597-L600).
@Gonals What are your thoughts on this?

[the remark this answers is not in the document]
Comment (janicduplessis) [reply src/pages/home/report/ReportActionsList.tsx:648]: Not sure if there is a bug with that logic but when testing comment linking I never see the loading indicator for newer messages, maybe I should make an issue for that?. Anyway I think it makes sense to always render the header / footer state if there's an error.

**Comment by janicduplessis** [comment]
@rojiphil I added the steps to test comment linking under QA Steps.

Addressed your review feedback too, let me know if it's good.

**Comment by mountiny** [comment]
@rojiphil What is your ETA for finishing this review? Can you please prioritize this one over others please?

**Comment by rojiphil** [comment]
> Addressed your review feedback too, let me know if it's good.

@janicduplessis Thanks for the changes. I just tested error condition for `GetNewerActions` and the error UI came up.
But we have another problem to address. When we retry `GetNewerActions`, the entire error message UI disappears for a moment and shows up again on failure of the API request. Like the `GetOlderActions`, shouldn't the error message Ui remain on retry until the API response is known? Here is a video to demonstrate this:

https://github.com/Expensify/App/assets/3004659/beeea6cf-9a18-4aaf-91ef-b4bdaaa8000f

**Comment by dannymcclain** [comment]
@rojiphil good catch. I think they should behave the same like you're suggesting.

**Comment by janicduplessis** [comment]
It currently disappears during loading because canShowHeader is false, which I think shouldn’t be. I’m a bit hesitant to add some workaround here since it seems like canShowHeader shouldn’t be false here and there might be a bug with it that could be addressed separately. What do you think?

**Review by rojiphil — COMMENTED** [review]
>  I’m a bit hesitant to add some workaround here

I agree that we should not implement any workaround around `canShowHeader`. But we may overcome the problem as mentioned in the review comments.
Would this work?

**Comment by janicduplessis** [comment]
I think it would work, but I am still hesitant on adding more complexity. For example we do rely on loading state in a few places like [here](https://github.com/janicduplessis/Expensify-App/blob/a251a37632f3258884dcc13fa1924a5b7329c615/src/pages/home/report/ReportActionsView.tsx#L354-L355) to not trigger additional requests. I am worried that if we have cases of loading data, but not setting this loading state might cause problems and make it more likely to introduce bugs in the future.

My current idea would be to merge it like this, and investigate the canShowHeader issue separately since currently there are just no loading states for new messages so it seems bugged. If canShowHeader is actually working as expected then maybe we can consider another solution.

**Comment by mountiny** [comment]
I agree we should avoid adding too much complexity. If it comes done to the issue with the button disappearing until new failure comes in, i think its fine for use to merge. Its rare case 

@rojiphil can you please continue with the checklist and treat that issues as nab?

**Review by rojiphil — APPROVED** [review]
LGTM and tests well.

**Comment by rojiphil** [comment]
> My current idea would be to merge it like this, and investigate the canShowHeader issue separately since currently there are just no loading states for new messages so it seems bugged. If canShowHeader is actually working as expected then maybe we can consider another solution.

Would be happy to work on this when we do.

**Comment by rojiphil** [comment]
> can you please continue with the checklist and treat that issues as nab?

@mountiny Checklist is done.

File: src/pages/home/report/ReportActionsView.tsx [hunk 5]
```diff
@@ -295,85 +305,118 @@ function ReportActionsView({
...
-            `[ReportActionsView] loadOlderChats ${JSON.stringify({
-                isOffline: network.isOffline,
-                isLoadingOlderReportActions,
-                isLoadingInitialReportActions,
-                oldestReportActionID: oldestReportAction?.reportActionID,
-                hasCreatedAction,
-                isTransactionThread: !isEmptyObject(transactionThreadReport),
-            })}`,
-        );
+    const loadOlderChats = useCallback(
```
Comment (mountiny) [inline src/pages/home/report/ReportActionsView.tsx:308]: I feel like these names here are misleading since we are not loading older or newer chats/ reports but only messages. NAB

File: src/styles/index.ts [hunk 6]
```diff
@@ -4494,14 +4494,23 @@ const styles = (theme: ThemeColors) =>
...
             bottom: 0,
             left: 0,
             right: 0,
             height: CONST.CHAT_HEADER_LOADER_HEIGHT,
         },
+        listBoundaryError: {
+            paddingVertical: 15,
+            paddingHorizontal: 20,
+        },
+        listBoundaryErrorText: {
```
Comment (mountiny) [inline src/styles/index.ts:4509]:
```suggestion

        listBoundaryErrorText: {
```

File: src/styles/index.ts [hunk 7]
```diff
@@ -4494,14 +4494,23 @@ const styles = (theme: ThemeColors) =>
...
-        chatBottomLoader: {
+        listBoundaryLoader: {
             position: 'absolute',
             top: 0,
             bottom: 0,
             left: 0,
             right: 0,
             height: CONST.CHAT_HEADER_LOADER_HEIGHT,
         },
+        listBoundaryError: {
```
Comment (mountiny) [inline src/styles/index.ts:4505]:
```suggestion

        listBoundaryError: {
```

**Review by mountiny — APPROVED** [review]
Just couple NABs, as this is useful performance / UX-wise, I am going to merge it as it, but I think we havent asked marketing for copy on the error so I will follow up with that

**Comment by janicduplessis** [comment]
Thanks @mountiny !

**Comment by kavimuru** [comment]
@janicduplessis Can you help us with this step? 

>  Test error state by failing pagination requests adding the following code in HttpUtils.ts [here](https://github.com/janicduplessis/Expensify-App/blob/16fa023201721525aadbd28e9975b0df53bddb80/src/libs/HttpUtils.ts#L108-L109). It might also be possible to simulate this by turning off internet after loading a chat.

**Comment by kavimuru** [comment]
@rojiphil @mountiny could you help with the QA steps.

> Test error state by failing pagination requests adding the following code in HttpUtils.ts [here](https://github.com/janicduplessis/Expensify-App/blob/16fa023201721525aadbd28e9975b0df53bddb80/src/libs/HttpUtils.ts#L108-L109). It might also be possible to simulate this by turning off internet after loading a chat.

**Comment by kavimuru** [comment]
@janicduplessis @rojiphil @mountiny Can we run only the following step?

> 1. Go in a long chat and scroll to an old message
> 2. Right click and "Copy link"
> 3. Go in another chat and paste the link
> 4. Clear the cache (Account Settings -> Troubleshoot -> Clear cache and restart)
> 5. Go back to the chat with the link and click it
> 6. You will be in the middle of a chat and can test loading older / newer messages

**Comment by janicduplessis** [comment]
Yea I think this might be hard to test on staging without being able to change the code. I will see if I can think of something.

**Comment by kavimuru** [comment]
@robertjchen @janicduplessis It would be great if you can validate this PR internally then.

**Comment by kavimuru** [comment]
@Beamanator Could this PR be validated internally?

**Comment by janicduplessis** [comment]
@rojiphil I figured out the issue that was preventing the loading state from showing, fix here https://github.com/Expensify/App/pull/42332

**Comment by Beamanator** [comment]
@mountiny @rojiphil @janicduplessis  can one of you please test this internally in staging? QA is having trouble 🙏

**Comment by mountiny** [comment]
@janicduplessis is that PR required to be CPed to staging or it can go through the normal deploy process?

I am not sure how this can be tested without changing code. I think we could however mark this off the checklist and we can monitor this change in the #newdot-quality room

**Comment by janicduplessis** [comment]
I think it can just go to normal deploy process. It is hard to test since it requires specific network failures or outage to happen. We can monitor it and make sure we do not have reports of infinite loading after network errors.

**Comment by aldo-expensify** [comment]
Getting old report actions gets stuck in a retry infinite loop, could it be related to this PR: https://expensify.slack.com/archives/C05LX9D6E07/p1717711999602319?thread_ts=1717462134.645329&cid=C05LX9D6E07

?
